File: packages/fleeter-daemon/src/ws/protocol.ts
/**
 * WebSocket protocol types for fleeter-daemon.
 */

// ─────────────────────────────────────────────────────────────────────────────
// Client Types
// ─────────────────────────────────────────────────────────────────────────────

export type ClientType = 'mcp' | 'tui';

export interface ConnectedClient {
  id: string;
  type: ClientType;
  connectedAt: Date;
  currentSessionId: string | null;
}

// ─────────────────────────────────────────────────────────────────────────────
// Client → Daemon Messages
// ─────────────────────────────────────────────────────────────────────────────

export interface ClientHello {
  type: 'hello';
  clientType: ClientType;
  clientId: string;
  version: string;
}

export interface CommandMessage {
  type: 'command';
  id: string;
  clientId: string;
  action: CommandAction;
  data?: Record<string, unknown>;
}

export type CommandAction =
  // Session management
  | 'create_session'
  | 'destroy_session'
  | 'list_sessions'
  | 'connect_session'
  | 'disconnect_session'
  // App lifecycle (Phase 3)
  | 'run_app'
  | 'stop_app'
  | 'hot_reload'
  | 'hot_restart'
  // Interaction tree (Phase 3)
  | 'get_tree'
  | 'execute_interaction'
  // Agent (Phase 4)
  | 'agent_message'
  // Status
  | 'get_status'
  | 'health_check';

export interface AgentToolCall {
  type: 'agent_tool_call';
  id: string;
  clientId: string;
  sessionId: string;
  intent: string;
  context?: string[];
  conversationId?: string;
}

export type ClientMessage = ClientHello | CommandMessage | AgentToolCall;

// ─────────────────────────────────────────────────────────────────────────────
// Daemon → Client Messages
// ─────────────────────────────────────────────────────────────────────────────

export interface ServerHello {
  type: 'hello_ack';
  daemonVersion: string;
  sessions: Array<{
    id: string;
    name: string;
    projectPath: string;
    appStatus: string;
  }>;
}

export interface CommandResponse {
  type: 'command_response';
  id: string;
  success: boolean;
  data?: unknown;
  error?: string;
}

export interface AgentResponse {
  type: 'agent_response';
  id: string;
  status: 'success' | 'failed' | 'needs_context';
  summary?: string;
  error?: string;
  question?: string;
  conversationId?: string;
}

export interface MonitoringEvent {
  type: 'event';
  ts: string;
  source: 'flutter' | 'vm' | 'agent' | 'daemon' | 'session';
  eventType: string;
  sessionId?: string;
  payload: unknown;
}

export type ServerMessage = ServerHello | CommandResponse | AgentResponse | MonitoringEvent;

// ─────────────────────────────────────────────────────────────────────────────
// Type Guards
// ─────────────────────────────────────────────────────────────────────────────

export function isClientHello(msg: unknown): msg is ClientHello {
  return typeof msg === 'object' && msg !== null && (msg as ClientHello).type === 'hello';
}

export function isCommandMessage(msg: unknown): msg is CommandMessage {
  return typeof msg === 'object' && msg !== null && (msg as CommandMessage).type === 'command';
}

export function isAgentToolCall(msg: unknown): msg is AgentToolCall {
  return typeof msg === 'object' && msg !== null && (msg as AgentToolCall).type === 'agent_tool_call';
}

File: tui/crates/agent-protocol/src/protocol.rs
//! WebSocket protocol types for communication with the daemon.

use serde::{Deserialize, Serialize};

// ─────────────────────────────────────────────────────────────────────────────
// Client Types
// ─────────────────────────────────────────────────────────────────────────────

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ClientType {
    Mcp,
    Tui,
}

// ─────────────────────────────────────────────────────────────────────────────
// Client → Daemon Messages
// ─────────────────────────────────────────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ClientHello {
    #[serde(rename = "type")]
    pub msg_type: String,
    #[serde(rename = "clientType")]
    pub client_type: ClientType,
    #[serde(rename = "clientId")]
    pub client_id: String,
    pub version: String,
}

impl ClientHello {
    pub const MSG_TYPE: &'static str = "hello";

    pub fn new(client_id: impl Into<String>, version: impl Into<String>) -> Self {
        Self {
            msg_type: Self::MSG_TYPE.to_string(),
            client_type: ClientType::Tui,
            client_id: client_id.into(),
            version: version.into(),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CommandMessage {
    #[serde(rename = "type")]
    pub msg_type: String,
    pub id: String,
    #[serde(rename = "clientId")]
    pub client_id: String,
    pub action: CommandAction,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub data: Option<serde_json::Value>,
}

impl CommandMessage {
    pub const MSG_TYPE: &'static str = "command";

    pub fn new(
        id: impl Into<String>,
        client_id: impl Into<String>,
        action: CommandAction,
    ) -> Self {
        Self {
            msg_type: Self::MSG_TYPE.to_string(),
            id: id.into(),
            client_id: client_id.into(),
            action,
            data: None,
        }
    }

    pub fn with_data(mut self, data: serde_json::Value) -> Self {
        self.data = Some(data);
        self
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum CommandAction {
    CreateSession,
    DestroySession,
    ListSessions,
    ConnectSession,
    DisconnectSession,
    RunApp,
    StopApp,
    HotReload,
    HotRestart,
    GetTree,
    ExecuteInteraction,
    GetLogs,
    AgentMessage,
    GetStatus,
    HealthCheck,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentToolCall {
    #[serde(rename = "type")]
    pub msg_type: String,
    pub id: String,
    #[serde(rename = "clientId")]
    pub client_id: String,
    #[serde(rename = "sessionId")]
    pub session_id: String,
    pub intent: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub context: Option<Vec<String>>,
    #[serde(rename = "conversationId", skip_serializing_if = "Option::is_none")]
    pub conversation_id: Option<String>,
}

impl AgentToolCall {
    pub const MSG_TYPE: &'static str = "agent_tool_call";

    pub fn new(
        id: impl Into<String>,
        client_id: impl Into<String>,
        session_id: impl Into<String>,
        intent: impl Into<String>,
    ) -> Self {
        Self {
            msg_type: Self::MSG_TYPE.to_string(),
            id: id.into(),
            client_id: client_id.into(),
            session_id: session_id.into(),
            intent: intent.into(),
            context: None,
            conversation_id: None,
        }
    }

    pub fn with_context(mut self, context: Vec<String>) -> Self {
        self.context = Some(context);
        self
    }

    pub fn with_conversation_id(mut self, id: impl Into<String>) -> Self {
        self.conversation_id = Some(id.into());
        self
    }
}

/// Union of all client messages.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(untagged)]
pub enum ClientMessage {
    Hello(ClientHello),
    Command(CommandMessage),
    AgentToolCall(AgentToolCall),
}

// ─────────────────────────────────────────────────────────────────────────────
// Daemon → Client Messages
// ─────────────────────────────────────────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ServerHello {
    #[serde(rename = "type")]
    pub msg_type: String,
    #[serde(rename = "daemonVersion")]
    pub daemon_version: String,
    pub sessions: Vec<SessionSummary>,
}

impl ServerHello {
    pub const MSG_TYPE: &'static str = "hello_ack";
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SessionSummary {
    pub id: String,
    pub name: String,
    #[serde(rename = "projectPath")]
    pub project_path: String,
    #[serde(rename = "appStatus")]
    pub app_status: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CommandResponse {
    #[serde(rename = "type")]
    pub msg_type: String,
    pub id: String,
    pub success: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub data: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

impl CommandResponse {
    pub const MSG_TYPE: &'static str = "command_response";
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum AgentResponseStatus {
    Success,
    Failed,
    NeedsContext,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentResponse {
    #[serde(rename = "type")]
    pub msg_type: String,
    pub id: String,
    pub status: AgentResponseStatus,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub summary: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub question: Option<String>,
    #[serde(rename = "conversationId", skip_serializing_if = "Option::is_none")]
    pub conversation_id: Option<String>,
}

impl AgentResponse {
    pub const MSG_TYPE: &'static str = "agent_response";
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum EventSource {
    Flutter,
    Vm,
    Agent,
    Daemon,
    Session,
    Tree,
    Interaction,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MonitoringEvent {
    #[serde(rename = "type")]
    pub msg_type: String,
    pub ts: String,
    pub source: EventSource,
    #[serde(rename = "eventType")]
    pub event_type: String,
    #[serde(rename = "sessionId", skip_serializing_if = "Option::is_none")]
    pub session_id: Option<String>,
    pub payload: serde_json::Value,
}

impl MonitoringEvent {
    pub const MSG_TYPE: &'static str = "event";
}

/// Union of all server messages.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum ServerMessage {
    #[serde(rename = "hello_ack")]
    Hello(ServerHelloData),
    #[serde(rename = "command_response")]
    CommandResponse(CommandResponseData),
    #[serde(rename = "agent_response")]
    AgentResponse(AgentResponseData),
    #[serde(rename = "event")]
    MonitoringEvent(MonitoringEventData),
    #[serde(rename = "agent_stream")]
    AgentStream(AgentStreamData),
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ServerHelloData {
    #[serde(rename = "daemonVersion")]
    pub daemon_version: String,
    pub sessions: Vec<SessionSummary>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CommandResponseData {
    pub id: String,
    pub success: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub data: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentResponseData {
    pub id: String,
    pub status: AgentResponseStatus,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub summary: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub question: Option<String>,
    #[serde(rename = "conversationId", skip_serializing_if = "Option::is_none")]
    pub conversation_id: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MonitoringEventData {
    pub ts: String,
    pub source: EventSource,
    #[serde(rename = "eventType")]
    pub event_type: String,
    #[serde(rename = "sessionId", skip_serializing_if = "Option::is_none")]
    pub session_id: Option<String>,
    pub payload: serde_json::Value,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentStreamData {
    pub id: String,
    #[serde(rename = "sessionId")]
    pub session_id: String,
    pub event: super::events::AgentEventType,
}
